### PR TITLE
feat: limit calling defineAuth multiple times

### PR DESCRIPTION
**Pull request by bombguy**
## Problem

Enforce call count limit on defineAuth to prevent customer from accidentally creating multiple AuthFactories. When we detect customer calling defineAuth multiple times, we should throw descriptive error with resolution steps.

**Issue number, if available:**

## Changes

Split up integration test into seperate files 
**Corresponding docs PR, if applicable:**

## Checklist

- [X] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/backend-auth/src/factory.ts [hunk 1]
```diff
@@ -124,10 +125,33 @@ class AmplifyAuthGenerator implements ConstructContainerEntryGenerator {
...
-export const defineAuth = (
+export const baseDefineAuth = (
   props: AmplifyAuthProps
 ): ConstructFactory<BackendAuth> =>
   new AmplifyAuthFactory(props, new Error().stack);
+
+// override function name for context.
+Object.defineProperty(baseDefineAuth, 'name', { value: 'defineAuth' });
+
+export const defineAuth = limitCalls(baseDefineAuth, 1);
```
Comment (edwardfoyle) [inline packages/backend-auth/src/factory.ts:157]: This seems a bit over-engineered. Could we instead do
```ts
let callCount = 0;

export const defineAuth = (
  props: AmplifyAuthProps
): ConstructFactory<BackendAuth> => {
  callCount = callCount + 1;
  if (callCount > 1) {
    throw new Error(...)
  }
  return new AmplifyAuthFactory(props, new Error().stack); 
}
```

File: packages/backend-auth/src/factory.ts [hunk 2]
```diff
@@ -45,7 +46,8 @@ export type AmplifyAuthProps = Expand<
 /**
  * Singleton factory for AmplifyAuth that can be used in Amplify project files
  */
-class AmplifyAuthFactory implements ConstructFactory<BackendAuth> {
+export class AmplifyAuthFactory implements ConstructFactory<BackendAuth> {
```
Comment (bombguy) [inline packages/backend-auth/src/factory.ts:49]: Need to export this class to access call count for testing. 

Made sure not to export unnecessary code from `src/index.ts`: 
https://github.com/aws-amplify/amplify-backend/pull/1046/files#diff-a43553012e47a2b89d0a4aacfd2d9e7ec9358de0bafc35399392034ca5538500L1

Comment (edwardfoyle) [reply packages/backend-auth/src/factory.ts:49]: I would leave a note in the JS doc that this class is only exported for testing and should not be exported from the package

File: packages/backend-auth/src/factory.ts [hunk 3]
```diff
@@ -45,7 +46,8 @@ export type AmplifyAuthProps = Expand<
...
+  static factoryCount = 0;
```
Comment (edwardfoyle) [inline packages/backend-auth/src/factory.ts:50]: could also add a note here that this is public for testing only

**Review by edwardfoyle — COMMENTED** [review]
Functionally LGTM, just some comments on wording

**Review by rtpascual — APPROVED** [review]

**Review by edwardfoyle — APPROVED** [review]

**Review by sobolk — APPROVED** [review]

**Pull request merged**